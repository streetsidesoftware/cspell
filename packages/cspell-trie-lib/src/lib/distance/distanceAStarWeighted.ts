import { PairingHeap } from '../utils/PairingHeap';
import { WeightMap } from './weightedMaps';

/**
 * Calculate the edit distance between two words using an A* algorithm.
 *
 * Using basic weights, this algorithm has the same results as the Damerau-Levenshtein algorithm.
 */
export function distanceAStarWeighted(a: string, b: string, map: WeightMap, cost = 100): number {
    const aN = a.length;
    const bN = b.length;

    const candidates = new PairingHeap(compare);

    candidates.add({ ai: 0, bi: 0, c: 0 });

    function opSub(n: Node) {
        const { ai, bi, c } = n;
        if (ai < aN && bi < bN) {
            const cc = a[ai] === b[bi] ? c : c + cost;
            candidates.add({ ai: ai + 1, bi: bi + 1, c: cc });
        }
    }

    function opIns(n: Node) {
        const { ai, bi, c } = n;
        if (bi < bN) {
            candidates.add({ ai: ai, bi: bi + 1, c: c + cost });
        }
    }

    function opDel(n: Node) {
        const { ai, bi, c } = n;
        if (ai < aN) {
            candidates.add({ ai: ai + 1, bi: bi, c: c + cost });
        }
    }

    function opSwap(n: Node) {
        const { ai, bi, c } = n;
        if (a[ai] === b[bi + 1] && a[ai + 1] === b[bi]) {
            candidates.add({ ai: ai + 2, bi: bi + 2, c: c + cost });
        }
    }

    function opMap(n: Node) {
        const { ai, bi, c } = n;
        const pos = { a, b, ai, bi, c };
        const costCalculations = [map.calcInsDelCosts(pos), map.calcSwapCosts(pos), map.calcReplaceCosts(pos)];
        costCalculations.forEach((iter) => {
            for (const nn of iter) {
                candidates.add(nn);
            }
        });
    }

    let best: Node | undefined;
    // const bc2 = 2 * bc;
    while ((best = candidates.dequeue())) {
        if (best.ai === aN && best.bi === bN) break;

        opSwap(best);
        opIns(best);
        opDel(best);
        opMap(best);
        opSub(best);
    }

    // istanbul ignore else
    return best ? best.c : -1;
}

interface Pos {
    /** the offset in string `a` */
    ai: number;
    /** the offset in string `b` */
    bi: number;
}

interface Node extends Pos {
    /** the current cost */
    c: number;
}

function compare(a: Node, b: Node): number {
    // Choose lowest cost or farthest Manhattan distance.
    return a.c - b.c || b.ai + b.bi - a.ai - a.bi;
}
