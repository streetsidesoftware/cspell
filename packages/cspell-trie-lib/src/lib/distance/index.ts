export { createWeightedMap, editDistance, editDistanceWeighted } from './distance';
export type { WeightMap } from './distance';
