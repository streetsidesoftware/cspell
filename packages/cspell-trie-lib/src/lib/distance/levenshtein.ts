/**
 * Damerau–Levenshtein distance
 * [Damerau–Levenshtein distance - Wikipedia](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance)
 * @param a - first word
 * @param b - second word
 * @returns Distance value
 */
export function levenshteinDistance(a: string, b: string): number {
    // By prefixing with spaces, no out of bounds checks are necessary.
    const aa = '  ' + a;
    const bb = '  ' + b;

    const nA = a.length + 1;
    const nB = b.length + 1;

    const firstRow: number[] = [];
    for (let i = 0; i <= nA; ++i) {
        firstRow[i] = i;
    }

    const matrix = [firstRow, [1, ...firstRow], [2, 1, ...firstRow]];
    let ppRow = matrix[0];
    let pRow = matrix[1];

    for (let j = 2; j <= nB; ++j) {
        const row = matrix[j % 3];
        row[0] = pRow[0] + 1;
        row[1] = pRow[1] + 1;

        const bp = bb[j - 1];
        const bc = bb[j];

        let ap = aa[0];

        for (let i = 2, i1 = 1; i <= nA; i1 = i, ++i) {
            const ac = aa[i];
            const c = pRow[i1] + (ac == bc ? 0 : 1);
            const ct = ac == bp && ap == bc ? ppRow[i1 - 1] + 1 : c;
            row[i] = Math.min(
                c, // substitute
                ct, // transpose
                pRow[i] + 1, // insert
                row[i1] + 1 // delete
            );
            ap = ac;
        }
        ppRow = pRow;
        pRow = row;
    }

    return pRow[nA];
}
