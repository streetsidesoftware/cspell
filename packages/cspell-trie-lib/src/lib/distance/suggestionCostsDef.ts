export type { SuggestionCostMapDef } from '@cspell/cspell-types';
