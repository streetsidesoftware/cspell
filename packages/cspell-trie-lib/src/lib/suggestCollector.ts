/**
 * This file is here to support code the referenced suggest directly and limit the exports.
 */
export {
    compSuggestionResults,
    defaultSuggestionCollectorOptions,
    impersonateCollector,
    isSuggestionResult,
    suggestionCollector,
} from './suggestions/suggestCollector';
export type {
    Cost,
    FilterWordFn,
    GenerateNextParam,
    GenerateSuggestionResult,
    MaxCost,
    Progress,
    SuggestionCollector,
    SuggestionCollectorOptions,
    SuggestionGenerator,
    SuggestionResult,
} from './suggestions/suggestCollector';
