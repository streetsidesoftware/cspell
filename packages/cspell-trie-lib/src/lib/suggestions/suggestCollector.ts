import { editDistanceWeighted, WeightMap } from '..';
import { createTimer } from '../utils/timer';
import { JOIN_SEPARATOR, WORD_SEPARATOR } from '../walker';

const defaultMaxNumberSuggestions = 10;

const BASE_COST = 100;
const MAX_NUM_CHANGES = 5;
const MAX_COST_SCALE = 0.5;
// max allowed cost scale should be a bit over 50% to allow for suggestions to short words, but not too high to have too many suggestions.
const MAX_ALLOWED_COST_SCALE = 1.03 * MAX_COST_SCALE;

const collator = new Intl.Collator();

const regexSeparator = new RegExp(`[${regexQuote(JOIN_SEPARATOR + WORD_SEPARATOR)}]`, 'g');

const wordLengthCost = [0, 50, 25, 5, 0];
const EXTRA_WORD_COST = 5;

/** time in ms */
const DEFAULT_COLLECTOR_TIMEOUT = 1000;

export type Cost = number;
export type MaxCost = Cost;

export interface SuggestionResult {
    word: string;
    cost: Cost;
}

export interface Progress {
    type: 'progress';
    /** Number of Completed Tasks so far */
    completed: number;
    /**
     * Number of tasks remaining, this number is allowed to increase over time since
     * completed tasks can generate new tasks.
     */
    remaining: number;
}

const symStopProcessing = Symbol('Collector Stop Processing');

export type GenerateNextParam = MaxCost | symbol | undefined;
export type GenerateSuggestionResult = SuggestionResult | Progress | undefined;

/**
 * Ask for the next result.
 * maxCost - sets the max cost for following suggestions
 * This is used to limit which suggestions are emitted.
 * If the `iterator.next()` returns `undefined`, it is to request a value for maxCost.
 *
 * The SuggestionIterator is generally the
 */
export type SuggestionGenerator = Generator<GenerateSuggestionResult, void, GenerateNextParam>;

// comparison function for Suggestion Results.
export function compSuggestionResults(a: SuggestionResult, b: SuggestionResult): number {
    return a.cost - b.cost || a.word.length - b.word.length || collator.compare(a.word, b.word);
}

export type FilterWordFn = (word: string, cost: number) => boolean;

export interface SuggestionCollector {
    /**
     * Collection suggestions from a SuggestionIterator
     * @param src - the SuggestionIterator used to generate suggestions.
     * @param timeout - the amount of time in milliseconds to allow for suggestions.
     * before sending `symbolStopProcessing`
     * Iterator implementation:
     * @example
     * r = yield(suggestion);
     * if (r === collector.symbolStopProcessing) // ...stop generating suggestions.
     */
    collect: (src: SuggestionGenerator, timeout?: number, filter?: FilterWordFn) => void;
    add: (suggestion: SuggestionResult) => SuggestionCollector;
    readonly suggestions: SuggestionResult[];
    readonly changeLimit: number;
    readonly maxCost: number;
    readonly word: string;
    readonly maxNumSuggestions: number;
    readonly includesTies: boolean;
    readonly ignoreCase: boolean;
    /**
     * Possible value sent to the SuggestionIterator telling it to stop processing.
     */
    readonly symbolStopProcessing: symbol;
}

export interface SuggestionCollectorOptions {
    /**
     * number of best matching suggestions.
     * @default 10
     */
    numSuggestions: number;

    /**
     * An optional filter function that can be used to limit remove unwanted suggestions.
     * I.E. to remove forbidden terms.
     * @default () => true
     */
    filter?: FilterWordFn;

    /**
     * The number of letters that can be changed when looking for a match
     * @default 5
     */
    changeLimit: number | undefined;

    /**
     * Include suggestions with tied cost even if the number is greater than `numSuggestions`.
     * @default true
     */
    includeTies?: boolean;

    /**
     * specify if case / accents should be ignored when looking for suggestions.
     * @default true
     */
    ignoreCase: boolean | undefined;

    /**
     * the total amount of time to allow for suggestions.
     * @default 1000
     */
    timeout?: number | undefined;

    /**
     * Used to improve the sorted results.
     */
    weightMap?: WeightMap | undefined;
}

export const defaultSuggestionCollectorOptions: SuggestionCollectorOptions = Object.freeze({
    numSuggestions: defaultMaxNumberSuggestions,
    filter: () => true,
    changeLimit: MAX_NUM_CHANGES,
    includeTies: false,
    ignoreCase: true,
    timeout: DEFAULT_COLLECTOR_TIMEOUT,
});

export function suggestionCollector(wordToMatch: string, options: SuggestionCollectorOptions): SuggestionCollector {
    const {
        filter = () => true,
        changeLimit = MAX_NUM_CHANGES,
        includeTies = false,
        ignoreCase = true,
        timeout = DEFAULT_COLLECTOR_TIMEOUT,
        weightMap,
    } = options;
    const numSuggestions = Math.max(options.numSuggestions, 0) || 0;
    const sugs = new Map<string, SuggestionResult>();
    let maxCost: number = BASE_COST * Math.min(wordToMatch.length * MAX_ALLOWED_COST_SCALE, changeLimit);

    let timeRemaining = timeout;

    function dropMax() {
        if (sugs.size < 2 || !numSuggestions) {
            sugs.clear();
            return;
        }
        const sorted = [...sugs.values()].sort(compSuggestionResults);
        let i = numSuggestions - 1;
        maxCost = sorted[i].cost;
        for (; i < sorted.length && sorted[i].cost <= maxCost; ++i) {
            /* empty */
        }
        for (; i < sorted.length; ++i) {
            sugs.delete(sorted[i].word);
        }
    }

    function adjustCost(sug: SuggestionResult): SuggestionResult {
        const words = sug.word.split(regexSeparator);
        const extraCost =
            words.map((w) => wordLengthCost[w.length] || 0).reduce((a, b) => a + b, 0) +
            (words.length - 1) * EXTRA_WORD_COST;
        return { word: sug.word, cost: sug.cost + extraCost };
    }

    function handleProgress(_progress: Progress) {
        // Do nothing.
    }

    function collectSuggestion(suggestion: SuggestionResult): MaxCost {
        const { word, cost } = adjustCost(suggestion);
        if (cost <= maxCost && filter(suggestion.word, cost)) {
            const known = sugs.get(word);
            if (known) {
                known.cost = Math.min(known.cost, cost);
            } else {
                sugs.set(word, { word, cost });
                if (cost < maxCost && sugs.size > numSuggestions) {
                    dropMax();
                }
            }
        }
        return maxCost;
    }

    /**
     * Collection suggestions from a SuggestionIterator
     * @param src - the SuggestionIterator used to generate suggestions.
     * @param timeout - the amount of time in milliseconds to allow for suggestions.
     */
    function collect(src: SuggestionGenerator, timeout?: number, filter?: FilterWordFn) {
        let stop: false | symbol = false;
        timeout = timeout ?? timeRemaining;
        timeout = Math.min(timeout, timeRemaining);
        if (timeout < 0) return;

        const timer = createTimer();

        let ir: IteratorResult<SuggestionResult | Progress | undefined>;
        while (!(ir = src.next(stop || maxCost)).done) {
            if (timer.elapsed() > timeout) {
                stop = symStopProcessing;
            }
            const { value } = ir;
            if (!value) continue;
            if (isSuggestionResult(value)) {
                if (!filter || filter(value.word, value.cost)) {
                    collectSuggestion(value);
                }
                continue;
            }
            handleProgress(value);
        }

        timeRemaining -= timer.elapsed();
    }

    function suggestions() {
        const rawValues = [...sugs.values()];
        const values = weightMap
            ? rawValues.map(({ word }) => ({ word, cost: editDistanceWeighted(wordToMatch, word, weightMap) }))
            : rawValues;

        const sorted = values.sort(compSuggestionResults);
        if (!includeTies && sorted.length > numSuggestions) {
            sorted.length = numSuggestions;
        }
        return sorted;
    }

    const collector: SuggestionCollector = {
        collect,
        add: function (suggestion: SuggestionResult) {
            collectSuggestion(suggestion);
            return this;
        },
        get suggestions() {
            return suggestions();
        },
        get maxCost() {
            return maxCost;
        },
        get word() {
            return wordToMatch;
        },
        get maxNumSuggestions() {
            return numSuggestions;
        },
        get changeLimit() {
            return changeLimit;
        },
        includesTies: includeTies,
        ignoreCase,
        symbolStopProcessing: symStopProcessing,
    };

    return collector;
}

export function isSuggestionResult(s: GenerateSuggestionResult): s is SuggestionResult {
    const r = s as Partial<SuggestionResult> | undefined;
    return r?.cost !== undefined && r.word != undefined;
}

/**
 *
 * @param text verbatim text to be inserted into a regexp
 * @returns text that can be used in a regexp.
 */
function regexQuote(text: string): string {
    return text.replace(/[[\]\-+(){},|*.\\]/g, '\\$1');
}
