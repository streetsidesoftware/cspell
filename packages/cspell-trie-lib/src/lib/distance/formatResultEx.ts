import type { ExResult } from './distanceAStarWeighted';
import { distanceAStarWeightedEx } from './distanceAStarWeighted';
import type { WeightMap } from './weightedMaps';

function pL(s: string, w: number) {
    const strWidth = vizWidth(s);
    const w0 = Math.max(0, w - strWidth);
    return ' '.repeat(w0) + s;
}

function pR(s: string, w: number) {
    const strWidth = vizWidth(s);
    const w0 = Math.max(0, w - strWidth);
    return s + ' '.repeat(w0);
}

function vizWidth(s: string) {
    const r = s.replace(/[\u0300-\u036F\u007f-\u009f]/gu, '');
    let i = 0;
    for (const c of r) {
        i += c.length;
    }
    return i;
}

export function formatExResult(ex: ExResult | undefined): string {
    if (!ex) return '<undefined>';

    const { cost, segments, penalty } = ex;
    const asString = segments.map(({ a, b, c, p }) => ({
        a: `<${a}>`,
        b: `<${b}>`,
        c: c.toString(10),
        p: p.toString(10),
    }));
    asString.push({
        a: '',
        b: '',
        c: ' = ' + segments.reduce((sum, { c }) => sum + c, 0).toString(10),
        p: ' = ' + segments.reduce((sum, { p }) => sum + p, 0).toString(10),
    });
    const parts = asString.map(({ a, b, c, p }) => ({
        a,
        b,
        c,
        p,
        w: Math.max(vizWidth(a), vizWidth(b), vizWidth(c), vizWidth(p)),
    }));
    const a = 'a: |' + parts.map(({ a, w }) => pR(a, w)).join('|') + '|';
    const b = 'b: |' + parts.map(({ b, w }) => pR(b, w)).join('|') + '|';
    const c = 'c: |' + parts.map(({ c, w }) => pL(c, w)).join('|') + '|';
    const p = 'p: |' + parts.map(({ p, w }) => pL(p, w)).join('|') + '|';
    const penaltyMsg = penalty ? `[+${penalty}]` : '';
    return `<${ex.a.slice(1, -1)}> -> <${ex.b.slice(1, -1)}> (${cost - penalty}${penaltyMsg})\n${[a, b, c, p].join(
        '\n'
    )}\n`;
}

export function formattedDistance(wordA: string, wordB: string, weightMap: WeightMap, cost?: number) {
    const x = distanceAStarWeightedEx(wordA, wordB, weightMap, cost);
    return formatExResult(x);
}
