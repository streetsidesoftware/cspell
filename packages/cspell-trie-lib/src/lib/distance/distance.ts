import { distanceAStarWeighted } from './distanceAStarWeighted';
import { levenshteinDistance } from './levenshtein';
import type { SuggestionCostMapDef } from './suggestionCostsDef';
import type { WeightMap } from './weightedMaps';
import { addDefToWeightMap, createWeightMap } from './weightedMaps';

export type { SuggestionCostMapDef } from './suggestionCostsDef';
export type { WeightMap } from './weightedMaps';

const defaultCost = 100;

/**
 * Calculate the edit distance between any two words.
 * Use the Damerauâ€“Levenshtein distance algorithm.
 * @param wordA
 * @param wordB
 * @param editCost - the cost of each edit (defaults to 100)
 * @returns the edit distance.
 */
export function editDistance(wordA: string, wordB: string, editCost = defaultCost): number {
    return levenshteinDistance(wordA, wordB) * editCost;
}

/**
 * Calculate the weighted edit distance between any two words.
 * @param wordA
 * @param wordB
 * @param weights - the weights to use
 * @param editCost - the cost of each edit (defaults to 100)
 * @returns the edit distance
 */
export function editDistanceWeighted(wordA: string, wordB: string, weights: WeightMap, editCost = defaultCost): number {
    return distanceAStarWeighted(wordA, wordB, weights, editCost);
}

/**
 * Collect Map definitions into a single weighted map.
 * @param defs - list of definitions
 * @returns A Weighted Map to be used with distance calculations.
 */
export function createWeightedMap(defs: SuggestionCostMapDef[]): WeightMap {
    return createWeightMap(...defs);
}

/**
 * Update a WeightedMap with a WeightedMapDef
 * @param weightedMap - map to update
 * @param def - the definition to use
 */
export function updatedWeightedMap(weightedMap: WeightMap, def: SuggestionCostMapDef): void {
    addDefToWeightMap(weightedMap, def);
}
