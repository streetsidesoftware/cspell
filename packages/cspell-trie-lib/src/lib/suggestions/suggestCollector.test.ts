import { createWeightedMap, editDistance, mapDictionaryInformationToWeightMap, WeightMap } from '..';
import { formattedDistance } from '../distance/formatResultEx';
import { suggestionCollector, SuggestionCollectorOptions, SuggestionGenerator } from './suggestCollector';

const defaultOptions: SuggestionCollectorOptions = {
    numSuggestions: 10,
    ignoreCase: undefined,
    changeLimit: undefined,
    timeout: undefined,
};

describe('Validate suggestCollector', () => {
    test('Tests the collector with filter', () => {
        const collector = suggestionCollector(
            'joyfully',
            sugOpts({ numSuggestions: 3, filter: (word) => word !== 'joyfully' })
        );
        collector.add({ word: 'joyfully', cost: 100 }).add({ word: 'joyful', cost: 100 });
        expect(collector.suggestions).toHaveLength(1);
    });

    test('Tests the collector with duplicate words of different costs', () => {
        const collector = suggestionCollector(
            'joyfully',
            sugOpts({ numSuggestions: 3, filter: (word) => word !== 'joyfully' })
        );
        collector.add({ word: 'joyful', cost: 100 });
        expect(collector.suggestions.length).toBe(1);
        collector.add({ word: 'joyful', cost: 75 });
        expect(collector.suggestions.length).toBe(1);
        expect(collector.suggestions[0].cost).toBe(75);
        collector
            .add({ word: 'joyfuller', cost: 200 })
            .add({ word: 'joyfullest', cost: 300 })
            .add({ word: 'joyfulness', cost: 340 })
            .add({ word: 'joyful', cost: 85 });
        expect(collector.suggestions.length).toBe(3);
        expect(collector.suggestions[0].cost).toBe(75);
    });

    test('collect', () => {
        const spellResults = [
            { word: 'joyfuller', cost: 200 },
            { word: 'joyfullest', cost: 300 },
            { word: 'joy', cost: 500 },
            { word: 'joyfulness', cost: 340 },
            { word: 'joyfuller', cost: 200 },
            { word: 'joy full', cost: 285 },
            { word: 'joyful', cost: 85 },
        ];
        const rValues: (number | undefined | symbol)[] = [];

        function* emit(): SuggestionGenerator {
            for (const r of spellResults) {
                rValues.push(yield r);
            }
        }

        const collector = suggestionCollector(
            'joyfully',
            sugOpts({ numSuggestions: 3, filter: (word) => word !== 'joyfully' })
        );
        collector.collect(emit());
        expect(collector.suggestions.length).toBe(3);
        expect(collector.suggestions[0].cost).toBe(85);
        // 285 -> 295 because the weight was adjusted because of the space.
        expect(rValues).toEqual([412, 412, 412, 412, 412, 300, 295]);
    });

    function s(word: string, cost: number) {
        return { word, cost };
    }

    // cspell:ignore joyo woudt
    test.each`
        word       | expected
        ${'word'}  | ${[s('word', 0), s('work', 100), s('words', 100)]}
        ${'words'} | ${[s('words', 0), s('word', 100), s('works', 100)]}
        ${'joy'}   | ${[s('joy', 5)]}
        ${'joyo'}  | ${[s('joy', 105), s('yo-yo', 200), s('joyous', 200)]}
        ${'woudt'} | ${[s('word', 200), s("won't", 200), s('words', 200), s('would', 200)]}
        ${'cafe'}  | ${[s('cafe', 0), s('café', 100), s('cafés', 200)]}
    `('collect suggestions for "$word"', ({ word, expected }) => {
        const collector = suggestionCollector(word, sugOpts({ numSuggestions: 3, changeLimit: 5, includeTies: true }));
        const sugs = sampleSuggestions().map((sugWord) => ({ word: sugWord, cost: editDistance(word, sugWord) }));
        sugs.forEach((sug) => collector.add(sug));
        expect(collector.suggestions).toEqual(expected);
    });

    // cspell:ignore aple
    test.each`
        word       | expected
        ${'word'}  | ${[s('word', 0), s('work', 110), s('words', 110)]}
        ${'words'} | ${[s('words', 0), s('word', 110), s('works', 110)]}
        ${'joy'}   | ${[s('joy', 0)]}
        ${'joyo'}  | ${[s('joy', 75), s('joyous', 165), s('yo-yo', 220)]}
        ${'woudt'} | ${[s('word', 220), s("won't", 220), s('words', 220), s('would', 220)]}
        ${'aple'}  | ${[s('apple', 55), s('apples', 165)]}
        ${'cafe'}  | ${[s('cafe', 0), s('café', 1), s('cafés', 111)]}
    `('collect weighted suggestions for "$word"', ({ word, expected }) => {
        const weightMap = sampleWeightMapNoDefaultWeights();
        const collector = suggestionCollector(
            word,
            sugOpts({ numSuggestions: 3, changeLimit: 5, includeTies: true, weightMap })
        );
        const sugs = sampleSuggestions().map((sugWord) => ({ word: sugWord, cost: editDistance(word, sugWord) }));
        sugs.forEach((sug) => collector.add(sug));

        const suggestions = collector.suggestions;
        suggestions
            .map((a) => a.word)
            .map((wordB) => formattedDistance(word.normalize('NFD'), wordB.normalize('NFD'), weightMap, 110))
            .forEach((r) => console.log(r));

        expect(suggestions).toEqual(expected);
    });

    // cspell:ignore aple
    test.each`
        word       | expected
        ${'word'}  | ${[s('word', 0), s('work', 100), s('words', 100)]}
        ${'words'} | ${[s('words', 0), s('word', 100), s('works', 100)]}
        ${'joy'}   | ${[s('joy', 0)]}
        ${'joyo'}  | ${[s('joy', 75), s('joyous', 155), s('yo-yo', 301)]}
        ${'woudt'} | ${[s('word', 200), s('words', 200), s('would', 200), s("won't", 210)]}
        ${'aple'}  | ${[s('apple', 55), s('apples', 155)]}
        ${'cafe'}  | ${[s('cafe', 0), s('café', 1), s('cafés', 101)]}
    `('collect weighted suggestions for "$word"', ({ word, expected }) => {
        const collector = suggestionCollector(
            word,
            sugOpts({ numSuggestions: 3, changeLimit: 5, includeTies: true, weightMap: sampleWeightMapDi() })
        );
        const sugs = sampleSuggestions().map((sugWord) => ({ word: sugWord, cost: editDistance(word, sugWord) }));
        sugs.forEach((sug) => collector.add(sug));
        expect(collector.suggestions).toEqual(expected);
    });
});

function sugOpts(opts: Partial<SuggestionCollectorOptions>): SuggestionCollectorOptions {
    return {
        ...defaultOptions,
        ...opts,
    };
}

function sampleSuggestions(): string[] {
    return ['']
        .concat(['joy', 'joyful', 'joyfully', 'joyous', 'enjoy', 'enjoyment', 'joyfulness', 'joyless', 'enjoys'])
        .concat(['one', 'two', 'concat', 'string', 'function', 'return', 'partial', 'values', 'value', 'collector'])
        .concat(['color', 'word', 'words', 'would', "wouldn't", "won't", 'water', 'walk', 'walking', 'cost'])
        .concat(['calculate', 'suggest', 'suggestion', 'supplement', 'apple', 'apples', 'walked', 'walker'])
        .concat(['yo-yo', 'the', 'saw', 'raw', 'paw', 'this', 'these', 'those', 'work', 'works', 'working'])
        .concat(['workable', 'worked', 'cafe', 'café', 'resume', 'résumé', 'cafés'])
        .concat([]);
}

function dictInfo() {
    return {
        locale: 'es-US',
        alphabet: 'a-zA-Z',
        suggestionEditCosts: [
            {
                map: 'aeiouy', // cspell:disable-line
                insDel: 75,
                replace: 50,
            },
            {
                map: "p(pp)|l(ll)|t(tt)|o(ou)(oh)|n(ing)('n)",
                replace: 55,
            },
            {
                // cspell:disable-next-line
                map: 'aàâäAÀÂÄ|eéèêëEÉÈÊË|iîïyIÎÏY|oôöOÔÖ|uùûüUÙÛÜ|cçCÇ|bB|dD|fF|gG|hH|jJ|kK|lL|mM|nN|pP|qQ|rR|sS|tT|vV|wW|xX|zZ',
                replace: 1,
            },
        ],
    };
}

function sampleWeightMapDi(): WeightMap {
    return mapDictionaryInformationToWeightMap(dictInfo());
}

function sampleWeightMapNoDefaultWeights(): WeightMap {
    return createWeightedMap(dictInfo().suggestionEditCosts);
}
