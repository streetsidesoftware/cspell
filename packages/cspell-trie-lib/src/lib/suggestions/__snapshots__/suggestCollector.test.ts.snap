// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate suggestCollector collect weighted suggestions for "aple" 1`] = `
Array [
  "<aple> -> <apple> (55)
a: |<^>|<a>|<p> |<l>|<e>|<$>|     |
b: |<^>|<a>|<pp>|<l>|<e>|<$>|     |
c: |  0|  0|  55|  0|  0|  0| = 55|
p: |  0|  0|   0|  0|  0|  0|  = 0|
",
  "<aple> -> <apples> (165)
a: |<^>|<a>|<p> |<l>|<e>|<> |<$>|      |
b: |<^>|<a>|<pp>|<l>|<e>|<s>|<$>|      |
c: |  0|  0|  55|  0|  0|110|  0| = 165|
p: |  0|  0|   0|  0|  0|  0|  0|   = 0|
",
]
`;

exports[`Validate suggestCollector collect weighted suggestions for "cafe" 1`] = `
Array [
  "<cafe> -> <cafe> (0)
a: |<^>|<c>|<a>|<f>|<e>|<$>|    |
b: |<^>|<c>|<a>|<f>|<e>|<$>|    |
c: |  0|  0|  0|  0|  0|  0| = 0|
p: |  0|  0|  0|  0|  0|  0| = 0|
",
  "<cafe> -> <cafe패> (1)
a: |<^>|<c>|<a>|<f>|<e>|<$>|    |
b: |<^>|<c>|<a>|<f>|<e패>|<$>|    |
c: |  0|  0|  0|  0|  1|  0| = 1|
p: |  0|  0|  0|  0|  0|  0| = 0|
",
  "<cafe> -> <cafe패s> (111)
a: |<^>|<c>|<a>|<f>|<e>|<> |<$>|      |
b: |<^>|<c>|<a>|<f>|<e패>|<s>|<$>|      |
c: |  0|  0|  0|  0|  1|110|  0| = 111|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
]
`;

exports[`Validate suggestCollector collect weighted suggestions for "joy" 1`] = `
Array [
  "<joy> -> <joy> (0)
a: |<^>|<j>|<o>|<y>|<$>|    |
b: |<^>|<j>|<o>|<y>|<$>|    |
c: |  0|  0|  0|  0|  0| = 0|
p: |  0|  0|  0|  0|  0| = 0|
",
]
`;

exports[`Validate suggestCollector collect weighted suggestions for "joyo" 1`] = `
Array [
  "<joyo> -> <joy> (75)
a: |<^>|<j>|<o>|<y>|<o>|<$>|     |
b: |<^>|<j>|<o>|<y>|<> |<$>|     |
c: |  0|  0|  0|  0| 75|  0| = 75|
p: |  0|  0|  0|  0|  0|  0|  = 0|
",
  "<joyo> -> <joyous> (165)
a: |<^>|<j>|<o>|<y>|<o> |<> |<$>|      |
b: |<^>|<j>|<o>|<y>|<ou>|<s>|<$>|      |
c: |  0|  0|  0|  0|  55|110|  0| = 165|
p: |  0|  0|  0|  0|   0|  0|  0|   = 0|
",
  "<joyo> -> <yo-yo> (220)
a: |<^>|<j>|<o>|<> |<y>|<o>|<$>|      |
b: |<^>|<y>|<o>|<->|<y>|<o>|<$>|      |
c: |  0|110|  0|110|  0|  0|  0| = 220|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
]
`;

exports[`Validate suggestCollector collect weighted suggestions for "word" 1`] = `
Array [
  "<word> -> <word> (0)
a: |<^>|<w>|<o>|<r>|<d>|<$>|    |
b: |<^>|<w>|<o>|<r>|<d>|<$>|    |
c: |  0|  0|  0|  0|  0|  0| = 0|
p: |  0|  0|  0|  0|  0|  0| = 0|
",
  "<word> -> <work> (110)
a: |<^>|<w>|<o>|<r>|<d>|<$>|      |
b: |<^>|<w>|<o>|<r>|<k>|<$>|      |
c: |  0|  0|  0|  0|110|  0| = 110|
p: |  0|  0|  0|  0|  0|  0|   = 0|
",
  "<word> -> <words> (110)
a: |<^>|<w>|<o>|<r>|<d>|<> |<$>|      |
b: |<^>|<w>|<o>|<r>|<d>|<s>|<$>|      |
c: |  0|  0|  0|  0|  0|110|  0| = 110|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
]
`;

exports[`Validate suggestCollector collect weighted suggestions for "words" 1`] = `
Array [
  "<words> -> <words> (0)
a: |<^>|<w>|<o>|<r>|<d>|<s>|<$>|    |
b: |<^>|<w>|<o>|<r>|<d>|<s>|<$>|    |
c: |  0|  0|  0|  0|  0|  0|  0| = 0|
p: |  0|  0|  0|  0|  0|  0|  0| = 0|
",
  "<words> -> <word> (110)
a: |<^>|<w>|<o>|<r>|<d>|<s>|<$>|      |
b: |<^>|<w>|<o>|<r>|<d>|<> |<$>|      |
c: |  0|  0|  0|  0|  0|110|  0| = 110|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
  "<words> -> <works> (110)
a: |<^>|<w>|<o>|<r>|<d>|<s>|<$>|      |
b: |<^>|<w>|<o>|<r>|<k>|<s>|<$>|      |
c: |  0|  0|  0|  0|110|  0|  0| = 110|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
]
`;

exports[`Validate suggestCollector collect weighted suggestions for "woudt" 1`] = `
Array [
  "<woudt> -> <word> (220)
a: |<^>|<w>|<o>|<u>|<d>|<t>|<$>|      |
b: |<^>|<w>|<o>|<r>|<d>|<> |<$>|      |
c: |  0|  0|  0|110|  0|110|  0| = 220|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
  "<woudt> -> <won't> (220)
a: |<^>|<w>|<o>|<u>|<d>|<t>|<$>|      |
b: |<^>|<w>|<o>|<n>|<'>|<t>|<$>|      |
c: |  0|  0|  0|110|110|  0|  0| = 220|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
  "<woudt> -> <words> (220)
a: |<^>|<w>|<o>|<u>|<d>|<t>|<$>|      |
b: |<^>|<w>|<o>|<r>|<d>|<s>|<$>|      |
c: |  0|  0|  0|110|  0|110|  0| = 220|
p: |  0|  0|  0|  0|  0|  0|  0|   = 0|
",
  "<woudt> -> <would> (220)
a: |<^>|<w>|<o>|<u>|<> |<d>|<t>|<$>|      |
b: |<^>|<w>|<o>|<u>|<l>|<d>|<> |<$>|      |
c: |  0|  0|  0|  0|110|  0|110|  0| = 220|
p: |  0|  0|  0|  0|  0|  0|  0|  0|   = 0|
",
]
`;
